### PR TITLE
Support glTF progress

### PR DESCRIPTION
**Pull request by zhuxudong**
### Please check if the PR fulfills these requirements
support #1651 
- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### Other information:

### CODE REVIEW

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]
Example crash

File: packages/loader/src/gltf/parser/GLTFParserContext.ts [hunk 1]
```diff
@@ -142,14 +149,17 @@ export class GLTFParserContext {
   ): void {
     const glTFResourceKey = glTFResourceMap[type];
     if (!glTFResourceKey) return;
+    this._taskCount++;
 
     if (type === GLTFParserType.Entity) {
       (this.glTFResource[glTFResourceKey] ||= [])[index] = <Entity>resource;
+      this._increaseProgress();
```
Comment (GuoLei1990) [inline packages/loader/src/gltf/parser/GLTFParserContext.ts:156]: Maybe Entity don't need

File: packages/loader/src/gltf/parser/GLTFParserContext.ts [hunk 2]
```diff
@@ -168,6 +178,14 @@ export class GLTFParserContext {
       });
     }
   }
+
+  private _increaseProgress() {
+    const progress = ++this._finishedTaskCount / this._taskCount;
+    if (progress > this._progress && progress < 1) {
+      this._progress = progress;
```
Comment (GuoLei1990) [inline packages/loader/src/gltf/parser/GLTFParserContext.ts:185]: You should write project code, because your `this._taskCount` is dynamic

**Review by GuoLei1990 — APPROVED** [review]

**Comment by GuoLei1990** [comment]
First，You can ignore synchronized things, such as entity
Also you didn’t consider schema

**Comment by GuoLei1990** [comment]
glTF:  progress = (json[D]+buffer[D]+texture[D+L]+animation[L]+skin[L]+scene[L])/ totalCount
glb:    progress = (buffer[D]+texture[L]+animation[L]+skin[L]+scene[L])/ totalCount

File: packages/design/src/IProgress.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Progress interface.
+ */
+export interface IProgress {
```
Comment (GuoLei1990) [inline packages/design/src/IProgress.ts:4]: Move to `core/src/asset`

File: packages/loader/src/gltf/parser/GLTFParserContext.ts [hunk 4]
```diff
@@ -107,9 +115,34 @@ export class GLTFParserContext {
...
     });
+
+    this._addProgressEvent(undefined, promise);
+    return promise;
+  }
+
+  /**
+   * @internal
+   */
+  _addProgressEvent(progress?: IProgress, taskPromise?: Promise<any>): void {
```
Comment (GuoLei1990) [inline packages/loader/src/gltf/parser/GLTFParserContext.ts:130]: _dispatchProgressEvent

File: packages/loader/src/gltf/parser/GLTFParserContext.ts [hunk 5]
```diff
@@ -107,9 +115,34 @@ export class GLTFParserContext {
...
+        this._progress.task.loaded += 1;
+        this._setProgress(this._progress);
+      });
+    }
+
+    if (progress) {
+      this._progress.detail = {
+        ...this._progress.detail,
+        ...progress.detail
+      };
```
Comment (GuoLei1990) [inline packages/loader/src/gltf/parser/GLTFParserContext.ts:143]: have GC

File: packages/design/src/IProgress.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  detail?: {
```
Comment (GuoLei1990) [inline packages/design/src/IProgress.ts:5]: Add comments

File: packages/loader/src/gltf/parser/GLTFParserContext.ts (hunk 4 above)
Comment (zhuxudong) [reply packages/loader/src/gltf/parser/GLTFParserContext.ts:130]: done

File: packages/loader/src/gltf/parser/GLTFParserContext.ts (hunk 5 above)
Comment (zhuxudong) [reply packages/loader/src/gltf/parser/GLTFParserContext.ts:143]: done

File: package.json [hunk 7]
```diff
@@ -65,5 +65,6 @@
       "eslint --fix",
       "git add"
     ]
-  }
-}
+  },
+  "repository": "git@github.com:galacean/engine.git"
```
[the remark this answers is not in the document]
Comment (zhuxudong) [reply package.json:69]: done

File: package.json [hunk 8]
```diff
@@ -65,5 +65,6 @@
       "eslint --fix",
       "git add"
     ]
-  }
-}
+  },
+  "repository": "git@github.com:galacean/runtime.git"
```
Comment (gz65555) [inline package.json:69]: Why add this line?

File: packages/design/src/IProgress.ts (hunk 3 above)
Comment (zhuxudong) [reply packages/design/src/IProgress.ts]: done

File: packages/design/src/IProgress.ts (hunk 6 above)
Comment (zhuxudong) [reply packages/design/src/IProgress.ts]: done

File: packages/core/src/asset/AssetPromise.ts [hunk 9]
```diff
@@ -173,3 +182,30 @@ enum PromiseState {
...
+ * Progress interface.
+ */
+export interface IProgress {
+  /**
+   * Progress detail.
+   * @remarks
+   * The key is the asset url, the value is the progress of the asset url.
+   * Some assets may have multiple urls, such as glTF, which has a glTF file and multiple texture files.
+   */
+  detail?: {
```
Comment (GuoLei1990) [inline packages/core/src/asset/AssetPromise.ts:196]: How do I direct which asset the current progress trigger is? Three and babylon have

File: package.json (hunk 8 above)
Comment (zhuxudong) [reply package.json:69]: Specify the place where your code lives. This is helpful for people who want to contribute. If the git repo is on GitHub, then the npm docs command will be able to find you.

File: packages/core/src/asset/AssetPromise.ts [hunk 10]
```diff
@@ -173,3 +182,30 @@ enum PromiseState {
...
 }
+
+/**
+ * Progress interface.
+ */
+export interface IProgress {
+  /**
+   * Progress detail.
+   * @remarks
+   * The key is the asset url, the value is the progress of the asset url.
```
Comment (GuoLei1990) [inline packages/core/src/asset/AssetPromise.ts:193]: The value is byte value?

File: package.json (hunk 8 above)
Comment (GuoLei1990) [reply package.json:69]: Our package is galacean/engine

File: packages/core/src/asset/AssetPromise.ts [hunk 11]
```diff
@@ -173,3 +182,30 @@ enum PromiseState {
   Rejected = "rejected",
   Canceled = "canceled"
 }
+
+/**
+ * Progress interface.
+ */
+export interface IProgress {
```
Comment (GuoLei1990) [inline packages/core/src/asset/AssetPromise.ts:189]:
```ts
export interface IProgress {
  ladingURL: string;
  requestDetails?: {
    [url: string]: {
      loaded: number;
      total: number;
    };
  };
  itemsLoaded: number;
  itemsTotal: number;
}
```

File: packages/loader/src/gltf/parser/GLTFParserContext.ts [hunk 12]
```diff
@@ -107,9 +115,31 @@ export class GLTFParserContext {
...
     });
+
+    this._dispatchProgressEvent(undefined, promise);
+    return promise;
+  }
+
+  /**
+   * @internal
+   */
+  _dispatchProgressEvent(progress?: IProgress, taskPromise?: Promise<any>): void {
```
Comment (gz65555) [inline packages/loader/src/gltf/parser/GLTFParserContext.ts:130]: This function should split two functions.

Comment (zhuxudong) [reply packages/loader/src/gltf/parser/GLTFParserContext.ts]: done

File: packages/core/src/asset/AssetPromise.ts (hunk 11 above)
Comment (zhuxudong) [reply packages/core/src/asset/AssetPromise.ts]: use two callback instead

File: packages/core/src/asset/AssetPromise.ts [hunk 13]
```diff
@@ -71,29 +74,44 @@ export class AssetPromise<T> implements PromiseLike<T> {
         if (this._state === PromiseState.Pending) {
           resolve(value);
           this._state = PromiseState.Fulfilled;
-          this._onProgressCallback = undefined;
+          this._onProgressItemsCallback = undefined;
+          this._onProgressDetailsCallback = undefined;
```
Comment (GuoLei1990) [inline packages/core/src/asset/AssetPromise.ts:78]: To prevent memory leaks?

File: packages/core/src/asset/AssetPromise.ts [hunk 14]
```diff
@@ -71,29 +74,44 @@ export class AssetPromise<T> implements PromiseLike<T> {
...
         }
       };
       const onCancel = (callback) => {
         if (this._state === PromiseState.Pending) {
           this._onCancelHandler = callback;
         }
       };
-      const setProgress = (progress: number) => {
+      const setItemsProgress = (loaded: number, total: number) => {
```
Comment (GuoLei1990) [inline packages/core/src/asset/AssetPromise.ts:94]: setItemsProgress should has `url` better

File: packages/core/src/asset/AssetPromise.ts [hunk 15]
```diff
@@ -102,12 +120,26 @@ export class AssetPromise<T> implements PromiseLike<T> {
    * @param callback
```
Comment (GuoLei1990) [inline packages/core/src/asset/AssetPromise.ts:120]: Fix comments

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]

File: packages/core/src/asset/AssetPromise.ts [hunk 16]
```diff
@@ -102,12 +120,26 @@ export class AssetPromise<T> implements PromiseLike<T> {
...
    * @returns AssetPromise
    */
-  onProgress(callback: (progress: number) => void): AssetPromise<T> {
-    if (this._progress) {
-      callback(this._progress);
+  onProgress(
+    itemsCallback?: (loaded: number, total: number) => void,
```
Comment (GuoLei1990) [inline packages/core/src/asset/AssetPromise.ts:124]: What is the value of the first optional parameter?

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]

File: packages/core/src/asset/AssetPromise.ts [hunk 17]
```diff
@@ -102,12 +120,26 @@ export class AssetPromise<T> implements PromiseLike<T> {
...
+    detailsCallback?: (url: string, loaded: number, total: number) => void
```
Comment (GuoLei1990) [inline packages/core/src/asset/AssetPromise.ts:125]: `detailsCallback` should changed to `detailCallback`?

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]

File: packages/core/src/asset/AssetPromise.ts (hunk 14 above)
Comment (zhuxudong) [reply packages/core/src/asset/AssetPromise.ts:94]: There are some common requirements such as `AssetPromise.all` does not require url

File: packages/core/src/asset/AssetPromise.ts (hunk 15 above)
Comment (zhuxudong) [reply packages/core/src/asset/AssetPromise.ts:120]: done

File: packages/loader/src/gltf/parser/GLTFParserContext.ts [hunk 18]
```diff
@@ -35,6 +35,15 @@ export class GLTFParserContext {
   buffers?: ArrayBuffer[];
 
   private _resourceCache = new Map<string, any>();
+  private _progress = {
+    detail: {},
+    task: { loaded: 0, total: 0 }
```
[the remark this answers is not in the document]
Comment (zhuxudong) [reply packages/loader/src/gltf/parser/GLTFParserContext.ts:40]: done

File: packages/core/src/asset/AssetPromise.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (zhuxudong) [reply packages/core/src/asset/AssetPromise.ts:124]: OK

File: packages/core/src/asset/AssetPromise.ts (hunk 17 above)
Comment (zhuxudong) [reply packages/core/src/asset/AssetPromise.ts:125]: OK，maybe `onTaskComplete`  and `onTaskDetail` is better

File: packages/core/src/asset/AssetPromise.ts (hunk 16 above)
Comment (zhuxudong) [reply packages/core/src/asset/AssetPromise.ts:124]: I have added detailed comments

**Review by GuoLei1990 — APPROVED** [review]

File: packages/core/src/asset/request.ts [hunk 19]
```diff
@@ -29,71 +29,63 @@ export type RequestConfig = {
...
     xhr.onload = () => {
       if (xhr.status < 200 || xhr.status >= 300) {
         reject(new Error(`request failed from: ${url}`));
         return;
       }
       const result = xhr.response ?? xhr.responseText;
-      resolve(result);
+      if (isImg) {
+        var blob = xhr.response;
+        var img = new Image();
```
Comment (gz65555) [inline packages/core/src/asset/request.ts:62]: use `const`

File: packages/core/src/asset/request.ts [hunk 20]
```diff
@@ -29,71 +29,63 @@ export type RequestConfig = {
...
+
     xhr.onload = () => {
       if (xhr.status < 200 || xhr.status >= 300) {
         reject(new Error(`request failed from: ${url}`));
         return;
       }
       const result = xhr.response ?? xhr.responseText;
-      resolve(result);
+      if (isImg) {
+        var blob = xhr.response;
```
Comment (gz65555) [inline packages/core/src/asset/request.ts:61]: You don't need cache blob variable

File: packages/core/src/asset/request.ts (hunk 19 above)
Comment (zhuxudong) [reply packages/core/src/asset/request.ts:62]: done

File: packages/core/src/asset/request.ts (hunk 20 above)
Comment (zhuxudong) [reply packages/core/src/asset/request.ts:61]: done

**Pull request merged**